src/natcap/invest/ndr/ndr_core.pyx

```diff
@@ -206,26 +227,32 @@ def ndr_eff_calculation(
 
                     working_retention_eff += (
                         intermediate_retention * neighbor.flow_proportion)
+                    neighbor = dn_iterator.next_no_skip()
 
                 if has_outflow:
+                    working_retention_eff /= dn_iterator.flow_dir_sum
                     effective_retention_raster.set(
                         global_col, global_row, working_retention_eff)
                 else:
                     raise Exception("got to a cell that has no outflow!")
             # search upslope to see if we need to push a cell on the stack
             # for i in range(8):
-            for neighbor in mfd_flow_direction_raster.get_upslope_neighbors(
-                    global_col, global_row):
+            up_iterator = UpslopeNeighborIterator(
+                mfd_flow_direction_raster, global_col, global_row)
+            neighbor = up_iterator.next()
+            while neighbor.direction < 8:
```

Comment:
I remember one of the things we talked about last week was the need to be careful to cover all of the iteration cases (e.g. `continue`, end of loop block) with calling `up_iterator.next()`.  What if we just handled that up here at the top of the loop? I suppose there's a slight cost to holding 2 neighbor objects in memory, but maybe it would be easier to maintain?  I'm thinking of something like the below (github wouldn't allow me to suggest something on the whole loop block)

```suggestion
            neighbor = up_iterator.next()
            next_neighbor = neighbor
            while neighbor.direction < 8:
                neighbor = next_neighbor
                next_neighbor = up_iterator.next()
```

Reply:
@phargogh I think I've found an even better solution which is to properly implement the iterators to meet the C++ iterator specification. That way we don't need to manually call `next()` at all. I was stumped on that before but I think I've worked it out.

Reply:
Woohoo!  That is super exciting.  Well done and I can't wait to see the iterator!